flink-ml-lib/src/main/java/org/apache/flink/ml/clustering/kmeans/OnlineKMeans.java

```diff
@@ -0,0 +1,569 @@
+                                    "ModelDataPartialUpdater",
+                                    new TupleTypeInfo<>(
+                                            TypeInformation.of(KMeansModelData.class),
+                                            DenseVectorTypeInfo.INSTANCE),
+                                    new ModelDataPartialUpdater(distanceMeasure, k))
+                            .setParallelism(parallelism)
+                            .connect(modelDataWithWeights.broadcast())
+                            .transform(
+                                    "ModelDataGlobalUpdater",
+                                    new TupleTypeInfo<>(
+                                            TypeInformation.of(KMeansModelData.class),
+                                            DenseVectorTypeInfo.INSTANCE),
+                                    new ModelDataGlobalUpdater(k, dims, parallelism, decayFactor))
+                            .setParallelism(1);
+
+            DataStream<KMeansModelData> outputModelData =
+                    modelDataWithWeights.map(
+                            (MapFunction<Tuple2<KMeansModelData, DenseVector>, KMeansModelData>)
+                                    x -> x.f0);
+
+            return new IterationBodyResult(
+                    DataStreamList.of(newModelDataWithWeights), DataStreamList.of(outputModelData));
+        }
+    }
+
+    private static class ModelDataGlobalUpdater
+            extends AbstractStreamOperator<Tuple2<KMeansModelData, DenseVector>>
+            implements TwoInputStreamOperator<
+                    Tuple2<KMeansModelData, DenseVector>,
+                    Tuple2<KMeansModelData, DenseVector>,
+                    Tuple2<KMeansModelData, DenseVector>> {
+        private final int k;
+        private final int dims;
+        private final int upstreamParallelism;
+        private final double decayFactor;
+
+        private ListState<Integer> partialModelDataReceivingState;
+        private ListState<Boolean> initModelDataReceivingState;
+        private ListState<KMeansModelData> modelDataState;
+        private ListState<DenseVector> weightsState;
+
+        private ModelDataGlobalUpdater(
+                int k, int dims, int upstreamParallelism, double decayFactor) {
+            this.k = k;
+            this.dims = dims;
+            this.upstreamParallelism = upstreamParallelism;
+            this.decayFactor = decayFactor;
+        }
+
+        @Override
+        public void initializeState(StateInitializationContext context) throws Exception {
+            super.initializeState(context);
+
+            partialModelDataReceivingState =
+                    context.getOperatorStateStore()
+                            .getListState(
+                                    new ListStateDescriptor<>(
+                                            "partialModelDataReceiving",
+                                            TypeInformation.of(Integer.class)));
```

Comment:
My understanding is that for basic types like int, we use `BasicTypeInfo.INT_TYPE_INFO`. Not sure if `TypeInformation.of(Integer.class)` would have inferior performance.

It seems simpler to just use `BasicTypeInfo.INT_TYPE_INFO`. Alternatively, please feel free to ask Yun Gao what is the recommended approach.

Reply:
I agree that `BasicTypeInfo.INT_TYPE_INFO` is a simpler choice. I'll make the change.